doobie/src/main/scala/tamer/db/package.scala

```diff
@@ -0,0 +1,111 @@
+package tamer
+
+import cats.effect.Blocker
+import com.sksamuel.avro4s.{Decoder, Encoder, SchemaFor}
+import doobie.Query0
+import doobie.hikari.HikariTransactor
+import doobie.implicits.{toDoobieStreamOps, _}
+import doobie.util.transactor.Transactor
+import eu.timepit.refined.auto._
+import fs2.Stream
+import log.effect.LogWriter
+import log.effect.zio.ZioLogWriter.log4sFromName
+import tamer.db.Compat.toIterable
+import tamer.db.ConfigDb.{DbConfig, QueryConfig}
+import tamer.db.Db.{Datable, TimeSegment, _}
+import tamer.kafka.Kafka
+import zio.blocking.Blocking
+import zio.interop.catz._
+import zio.{Queue, Task, UIO, ZIO, _}
+
+import java.sql.SQLException
+import java.time.{Duration, Instant}
+import scala.concurrent.ExecutionContext
+
+package object db {
+  type DbTransactor  = Has[Transactor[Task]]
+  type TamerDBConfig = DbTransactor with QueryConfig
+
+  private[this] implicit final class InstantOps(private val instant: Instant) extends AnyVal {
+    def orNow: UIO[Instant] =
+      UIO(Instant.now).map {
+        case now if instant.isAfter(now) => now
+        case _                           => instant
+      }
+  }
+
+  final def mkSetup[K <: Product: Encoder: Decoder: SchemaFor, V <: Product with Datable: Ordering: Encoder: Decoder: SchemaFor](
+      queryBuilder: TimeSegment => Query0[V]
+  )(earliest: Instant, tumblingStep: Duration, keyExtract: V => K): Setup[K, V, TimeSegment] = {
+
+    val timeSegment = TimeSegment(earliest, earliest.plus(tumblingStep))
+
+    val qBuilder = new QueryBuilder[V, TimeSegment] {
+      override val queryId: Int                         = queryBuilder(timeSegment).sql.hashCode
+      override def query(state: TimeSegment): Query0[V] = queryBuilder(state)
+    }
+
+    def stateFold(timeSegment: TimeSegment)(queryResult: QueryResult[V]): UIO[TimeSegment] =
+      if (queryResult.results.isEmpty) timeSegment.to.plus(tumblingStep).orNow.map(TimeSegment(timeSegment.from, _))
+      else {
+        val mostRecent = queryResult.results.max.instant
+        mostRecent.plus(tumblingStep).orNow.map(TimeSegment(mostRecent, _))
```

Comment:
I'm afraid that doing it here would prevent clients from doing things like: well, once I'm 2/3 seconds wall clock time from current time (yes, they are subject to skew so it's a big assumption) I'm going to back off and run queries slower (see https://github.com/laserdisc-io/tamer/issues/44#issuecomment-737384367)

I'm sure we could work around this but in general removing client control from what `State` makes sense to them and constrain it to be a `TimeSegment` might be counter productive (e.g. a client that always needs to pull the entire table or one that needs to scan through a sortable primary key). Alternatively, we provide a factory for multiple common cases (this would be one) and allow one that is much like the original (leaving the movement of the cursor completely to the caller)

Reply:
Yes this is the first part of the second scenario. I wanted to work from a specific practical case to establish a baseline. I’m going to build the more generic setup next and it will be feature parity. The reason is to optimize for (what I think will be) the common case, and then give the user freedom to write his own function if he likes.